File: lib/java/main/src/test/java/net/alphatab/alphaskia/MusicSheetRenderTest.java
package net.alphatab.alphaskia;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public abstract class MusicSheetRenderTest {
    private static AlphaSkiaTypeface _musicTypeface;
    private static final int _musicFontSize = 34;
    private static final float _renderScale = 1.0f;

    protected static float getRenderScale() {
        return _renderScale;
    }

    protected static AlphaSkiaTypeface getMusicTypeface() {
        return _musicTypeface;
    }

    protected static int getMusicFontSize() {
        return _musicFontSize;
    }

    private static final Map<String, AlphaSkiaTypeface> _customTypefaces =
            new HashMap<>();

    private static String customTypefaceKey(String fontFamily, boolean isBold, boolean isItalic) {
        return fontFamily.toLowerCase() + "_" + isBold + "_" + isItalic;
    }

    @BeforeAll
    public static void setup() throws IOException {
        var os = System.getProperty("os.name");
        String libName;
        String target;
        if (os.startsWith("Mac")) {
            target = "osx";
            libName = "libAlphaSkiaJni.dylib";
        } else if (os.startsWith("Win")) {
            target = "win";
            libName = "libAlphaSkiaJni.dll";
        } else if (os.startsWith("Linux")) {
            target = "linux";
            libName = "libAlphaSkiaJni.so";
        } else {
            target = os.toLowerCase();
            libName = "libAlphaSkiaJni";
        }

        var jarch = System.getProperty("os.arch");
        var arch = switch (jarch) {
            case "x86", "i368", "i486", "i586", "i686" -> "x86";
            case "x86_64", "amd64" -> "x64";
            case "arm" -> "arm";
            case "aarch64" -> "arm64";
            default -> jarch;
        };

        AlphaSkiaPlatform.loadLibrary(new String[]{
                Path.of(System.getProperty("alphaskia.library.path"), "libAlphaSkiaJni-" + target + "-" + arch + "-shared", libName).toAbsolutePath().toString()
        });

        var bravura = readFont("font/bravura/Bravura.ttf");

        _musicTypeface = AlphaSkiaTypeface.register(bravura);

        _customTypefaces.put(customTypefaceKey("Roboto", false, false),
                AlphaSkiaTypeface.register(readFont("font/roboto/Roboto-Regular.ttf")));
        _customTypefaces.put(customTypefaceKey("Roboto", true, false),
                AlphaSkiaTypeface.register(readFont("font/roboto/Roboto-Bold.ttf")));
        _customTypefaces.put(customTypefaceKey("Roboto", false, true),
                AlphaSkiaTypeface.register(readFont("font/roboto/Roboto-Italic.ttf")));
        _customTypefaces.put(customTypefaceKey("Roboto", true, true),
                AlphaSkiaTypeface.register(readFont("font/roboto/Roboto-BoldItalic.ttf")));

        _customTypefaces.put(customTypefaceKey("PT Serif", false, false),
                AlphaSkiaTypeface.register(readFont("font/ptserif/PTSerif-Regular.ttf")));
        _customTypefaces.put(customTypefaceKey("PT Serif", true, false),
                AlphaSkiaTypeface.register(readFont("font/ptserif/PTSerif-Bold.ttf")));
        _customTypefaces.put(customTypefaceKey("PT Serif", false, true),
                AlphaSkiaTypeface.register(readFont("font/ptserif/PTSerif-Italic.ttf")));
        _customTypefaces.put(customTypefaceKey("PT Serif", true, true),
                AlphaSkiaTypeface.register(readFont("font/ptserif/PTSerif-BoldItalic.ttf")));
    }

    protected static AlphaSkiaTypeface getTypeface(String name, boolean isBold, boolean isItalic) {
        var face = _customTypefaces.getOrDefault(customTypefaceKey(name, isBold, isItalic), null);
        if (face == null) {
            throw new IllegalStateException("Unknown font requested: " + name);
        }
        return face;
    }


    private AlphaSkiaTextAlign textAlign = AlphaSkiaTextAlign.CENTER;

    protected AlphaSkiaTextAlign getTextAlign() {
        return textAlign;
    }

    protected void setTextAlign(AlphaSkiaTextAlign textAlign) {
        this.textAlign = textAlign;
    }

    private AlphaSkiaTextBaseline textBaseline = AlphaSkiaTextBaseline.TOP;

    protected AlphaSkiaTextBaseline getTextBaseline() {
        return textBaseline;
    }

    protected void setTextBaseline(AlphaSkiaTextBaseline textBaseline) {
        this.textBaseline = textBaseline;
    }

    private AlphaSkiaTypeface typeface;

    protected AlphaSkiaTypeface getTypeface() {
        return typeface;
    }

    protected void setTypeface(AlphaSkiaTypeface typeface) {
        this.typeface = typeface;
    }

    private float fontSize;

    public float getFontSize() {
        return fontSize;
    }

    public void setFontSize(float fontSize) {
        this.fontSize = fontSize;
    }

    private static byte[] readFont(String relativePath) throws IOException {
        var fullPath = Path.of(System.getProperty("testdata.path")).resolve(relativePath);
        try (var stream = new FileInputStream(fullPath.toFile())) {
            return stream.readAllBytes();
        }
    }

    private AlphaSkiaImage renderFullImage() {
        try (var canvas = new AlphaSkiaCanvas()) {
            var images = Arrays.stream(getAllParts()).map(p -> p.render(canvas)).toArray(AlphaSkiaImage[]::new);

            canvas.beginRender(getTotalWidth(), getTotalHeight());

            var partPositions = getPartPositions();
            for (var i = 0; i < images.length; i++) {
                canvas.drawImage(images[i], partPositions[i][0], partPositions[i][1], images[i].getWidth() / getRenderScale(), images[i].getHeight() / getRenderScale());
                images[i].close();
            }

            return canvas.endRender();
        }
    }

    @Test
    public void renderTest() throws IOException {
        try (var finalImage = renderFullImage()) {
            var testOutputPath = Path.of(System.getProperty("testoutput.path"));
            Files.createDirectories(testOutputPath);

            var file = testOutputPath.resolve(this.getClass().getSimpleName() +".png").toFile();
            byte[] png = finalImage.toPng();

            try (var stream = new FileOutputStream(file)) {
                stream.write(png);
            }
        }

    }

    protected abstract int getTotalWidth();

    protected abstract int getTotalHeight();

    protected abstract float[][] getPartPositions();

    protected abstract RenderFunction[] getAllParts();
}

File: test/dotnet/AlphaSkia.Test_/MusicSheetRenderTest.cs
namespace AlphaSkia.Test;

public abstract class MusicSheetRenderTest
{
    protected static AlphaSkiaTypeface MusicTypeface { get; private set; } = null!;
    protected static float RenderScale { get; private set; } = 1;
    protected static int MusicFontSize { get; private set; } = 34;

    private static readonly IDictionary<string, AlphaSkiaTypeface> CustomTypefaces =
        new Dictionary<string, AlphaSkiaTypeface>(StringComparer.OrdinalIgnoreCase);


    private static string CustomTypefaceKey(string fontFamily, bool isBold, bool isItalic)
    {
        return fontFamily.ToLowerInvariant() + "_" + isBold + "_" + isItalic;
    }

    [OneTimeSetUp]
    public static void Setup()
    {
        AlphaSkiaPlatform.LoadLibrary();

        var bravura = ReadFont("font/bravura/Bravura.ttf");

        MusicTypeface = AlphaSkiaTypeface.Register(bravura) ??
                        throw new InvalidOperationException("Could not load bravura font");

        CustomTypefaces[CustomTypefaceKey("Roboto", false, false)] =
            AlphaSkiaTypeface.Register(ReadFont("font/roboto/Roboto-Regular.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("Roboto", true, false)] =
            AlphaSkiaTypeface.Register(ReadFont("font/roboto/Roboto-Bold.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("Roboto", false, true)] =
            AlphaSkiaTypeface.Register(ReadFont("font/roboto/Roboto-Italic.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("Roboto", true, true)] =
            AlphaSkiaTypeface.Register(ReadFont("font/roboto/Roboto-BoldItalic.ttf")) ??
            throw new InvalidOperationException("Could not load font");

        CustomTypefaces[CustomTypefaceKey("PT Serif", false, false)] =
            AlphaSkiaTypeface.Register(ReadFont("font/ptserif/PTSerif-Regular.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("PT Serif", true, false)] =
            AlphaSkiaTypeface.Register(ReadFont("font/ptserif/PTSerif-Bold.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("PT Serif", false, true)] =
            AlphaSkiaTypeface.Register(ReadFont("font/ptserif/PTSerif-Italic.ttf")) ??
            throw new InvalidOperationException("Could not load font");
        CustomTypefaces[CustomTypefaceKey("PT Serif", true, true)] =
            AlphaSkiaTypeface.Register(ReadFont("font/ptserif/PTSerif-BoldItalic.ttf")) ??
            throw new InvalidOperationException("Could not load font");
    }

    protected static AlphaSkiaTypeface GetTypeface(string name, bool isBold, bool isItalic)
    {
        if (!CustomTypefaces.TryGetValue(CustomTypefaceKey(name, isBold, isItalic), out var face))
        {
            throw new InvalidOperationException("Unknown font requested: " + name);
        }

        return face;
    }


    protected AlphaSkiaTextAlign TextAlign { get; set; } = AlphaSkiaTextAlign.Center;
    protected AlphaSkiaTextBaseline TextBaseline { get; set; } = AlphaSkiaTextBaseline.Top;
    protected AlphaSkiaTypeface Typeface { get; set; } = null!;
    protected float FontSize { get; set; }

    private static byte[] ReadFont(string relativePath)
    {
        var resourceName = "AlphaSkia.Test." + relativePath.Replace("/", ".");
        using var stream = typeof(MusicSheetRenderTest).Assembly.GetManifestResourceStream(resourceName);
        if (stream == null)
        {
            throw new IOException("Could not find resource " + relativePath);
        }

        var data = new byte[stream.Length];
        stream.ReadExactly(data);
        return data;
    }

    private AlphaSkiaImage RenderFullImage()
    {
        using var canvas = new AlphaSkiaCanvas();

        var images = AllParts.Select(p => p(canvas)).ToArray();

        canvas.BeginRender(TotalWidth, TotalHeight, RenderScale);

        for (var i = 0; i < images.Length; i++)
        {
            canvas.DrawImage(images[i], PartPositions[i, 0], PartPositions[i, 1], images[i].Width / RenderScale, images[i].Height / RenderScale);
            images[i].Dispose();
        }

        return canvas.EndRender()!;
    }

    [Test]
    public void RenderTest()
    {
        using var finalImage = RenderFullImage();
        var fileName = GetType().Name;
        File.WriteAllBytes(Path.Combine(TestContext.CurrentContext.WorkDirectory, fileName + ".png"),
            finalImage.ToPng()!);
    }

    protected abstract int TotalWidth { get; }
    protected abstract int TotalHeight { get; }
    protected abstract float[,] PartPositions { get; }

    protected abstract Func<AlphaSkiaCanvas, AlphaSkiaImage>[] AllParts { get; }
}